Library/Homebrew/cask/cask.rb

```diff
@@ -292,15 +292,13 @@ def tap_git_head
     def populate_from_api!(json_cask)
       raise ArgumentError, "Expected cask to be loaded from the API" unless loaded_from_api?
 
+      @caskfile_only = json_cask[:caskfile_only]
       @languages = json_cask.fetch(:languages, [])
       @tap_git_head = json_cask.fetch(:tap_git_head, "HEAD")
 
       @ruby_source_path = json_cask[:ruby_source_path]
-
-      # TODO: Clean this up when we deprecate the current JSON API and move to the internal JSON v3.
-      ruby_source_sha256 = json_cask.dig(:ruby_source_checksum, :sha256)
-      ruby_source_sha256 ||= json_cask[:ruby_source_sha256]
-      @ruby_source_checksum = { "sha256" => ruby_source_sha256 }
+      @ruby_source_checksum = json_cask[:ruby_source_checksum] || # public JSON v2
+                              json_cask[:ruby_source_sha256]&.then { { "sha256" => _1 } } # internal JSON v3
```

Comment:
I ended up removing this TODO because we won't be able to remove this going forward. We need to keep backwards compatibility with all of the JSON cask files that get stored when casks get installed from the API.

Reply:
> We need to keep backwards compatibility with all of the JSON cask files that get stored when casks get installed from the API.

@apainintheneck Oof. That seems undesirable (but clearly not your fault). Could we add some sort of versioning to these files/filenames so this isn't going to have to be the case indefinitely?

Reply:
I think we'll have to keep backwards compatibility until all of the currently supported macOS versions reach end of life. It will be a while. I'm not sure how we handle very long term deprecations like that.

Reply:
@apainintheneck That's not tenable given how recently this was introduced and that we have e.g. already two paths for this (writing formula files, writing JSON files). Instead, we should handle when this JSON file is in an invalid format and act the same as if it were missing entirely (which is something that already works).

Reply:
I guess that is another option. We don't even need to explicitly handle it being in an invalid format since as you already said we handle that already. I guess in that case it shouldn't be that big of a deal.

https://github.com/Homebrew/brew/blob/7b2bfee3638cba1571157f07fd447d9949bd0a6a/Library/Homebrew/cask/installer.rb#L576-L593

We already fall back to the current cask file whenever the old cask file can't be loaded. I still think backwards compatibility is cheap here though and I don't understand why we would want to break it. Breaking backwards compatibility here could potentially affect all casks installed in the last year from the API.

Reply:
If we really wanted to be backwards compatible, we could use the original idea of just removing unnecessary fields from the JSON blob without really changing it's structure. Then, we wouldn't have to worry too much about the persistent if/else logic in the cask loader since it would need fewer changes.

Reply:
> Breaking backwards compatibility here could potentially affect all casks installed in the last year from the API.

I am not proposing we break it immediately. I am saying that:

> I think we'll have to keep backwards compatibility until all of the currently supported macOS versions reach end of life.

Is _way_ too long to avoid breaking it and we with internal JSON v3 we should not add Yet Another Path That We'll Need To Maintain Compatibility For.

> If we really wanted to be backwards compatible, we could use the original idea of just removing unnecessary fields from the JSON blob without really changing it's structure.

For uninstall etc.: yeh, this is exactly what I'm thinking we should do.

We could also have e.g. something like `brew update` go through and cleanup existing JSON files (and even generate new JSON files from cask files) before we break backwards compatibility here.

Reply:
@MikeMcQuaid Is this your position right now?

1. It doesn't make sense to have a third file format like the internal JSON v3 representation of casks that gets stored when the cask is installed.
2. It doesn't make sense to support three cask file formats indefinitely.
3. Too much irrelevant information has caused problems when loading old cask files in the past.
4. A third slim file format that represents only the logic needed to uninstall a cask should be added.

Things I have to look into a bit more.
- Do we use the installed cask file for anything other than uninstalling casks?
- Is it possible to have an installed cask that is not in a tap and only references the installed cask file?
- Are there any parts of the uninstall process that require the cask file and don't work with the JSON representation?

Reply:
> @MikeMcQuaid Is this your position right now?
> 
> 1. It doesn't make sense to have a third file format like the internal JSON v3 representation of casks that gets stored when the cask is installed.
> 2. It doesn't make sense to support three cask file formats indefinitely.
> 3. Too much irrelevant information has caused problems when loading old cask files in the past.
> 4. A third slim file format that represents only the logic needed to uninstall a cask should be added.

Yes, I think so, thanks for clarifying 👍🏻 

Also perhaps 5. the third slim file format would ideally just be a subset (e.g. with `Hash#slice` of the internal JSON v3).

> * Do we use the installed cask file for anything other than uninstalling casks?
> * Are there any parts of the uninstall process that require the cask file and don't work with the JSON representation?

I don't think so but don't know for sure. @Rylan12 @reitermarkus @Homebrew/cask folks: help? #cask in Slack may help here too.

> * Is it possible to have an installed cask that is not in a tap and only references the installed cask file?

Yes, easy to reproduce too:
- install cask from tap
- untap tap
- uninstall cask

Reply:
> > * Is it possible to have an installed cask that is not in a tap and only references the installed cask file?
> 
> Yes, easy to reproduce too:
> 
> * install cask from tap
> * untap tap
> * uninstall cask

In that case, we definitely shouldn't do this. It means that you'd only be able to load a cask for any of the commands that require that if it's currently available in a tap. Anything that got removed at some point would not be usable with any of the other brew commands.

Reply:
Is the compatibility code really that bad? It's just three bits of branching logic. The easiest way to avoid that branching would be to have more conservative changes to the JSON layout as I proposed before in https://github.com/Homebrew/brew/pull/16514. That would remove the need for any compatibility code.

Reply:
I do like the idea of transforming the old JSON files into the expected format but at the same time that just seems like moving the compatibility code somewhere else in the codebase.

Reply:
> In that case, we definitely shouldn't do this.

What is "this", sorry?

> It means that you'd only be able to load a cask for any of the commands that require that if it's currently available in a tap.

This seems reasonable but I don't care enough to argue about it any more if you feel strongly about this.

The alternative is: write a JSON file into the Caskroom in the same format whether installing from homebrew-cask, other official taps or a 3rd-party tap. Again: the status quo of "install caskfile sometimes and JSON file other times" is not a good one.

> Is the compatibility code really that bad?

If it was all neatly sorted into methods or a class/classes named accordingly: no big deal. As-is: it's going to be hard to extract later.

> The easiest way to avoid that branching would be to have more conservative changes to the JSON layout

Again: the more conservative the approaches to changing the JSON, the less it seems worthy of having an internal JSON and/or a v3 JSON at all.

To be clear: I think the v3 JSON is very desirable and the internal one is: we just need to do it right. This is an API format that may still be with us in 5-10 years.

> I do like the idea of transforming the old JSON files into the expected format but at the same time that just seems like moving the compatibility code somewhere else in the codebase.

If it's in a single location that's easier to delete/remove/deprecate later: that's better than having it inline.

See the OS layer as an example: it makes it much easier than just having `if OS.mac?` scattered all over the place.

Reply:
> > In that case, we definitely shouldn't do this.
> 
> What is "this", sorry?

My bad. I was talking about the idea of having a reduced info cask JSON that could only be used to uninstall casks.

> > It means that you'd only be able to load a cask for any of the commands that require that if it's currently available in a tap.
> 
> This seems reasonable but I don't care enough to argue about it any more if you feel strongly about this.
> 
> The alternative is: write a JSON file into the Caskroom in the same format whether installing from homebrew-cask, other official taps or a 3rd-party tap. Again: the status quo of "install caskfile sometimes and JSON file other times" is not a good one.

The cask JSON doesn't support all of the features found in the cask Ruby file right now so I don't think we'll be able to do that unless we change how the Ruby file works.

> > Is the compatibility code really that bad?
> 
> If it was all neatly sorted into methods or a class/classes named accordingly: no big deal. As-is: it's going to be hard to extract later.

Fair point. 👍

> > The easiest way to avoid that branching would be to have more conservative changes to the JSON layout
> 
> Again: the more conservative the approaches to changing the JSON, the less it seems worthy of having an internal JSON and/or a v3 JSON at all.
> 
> To be clear: I think the v3 JSON is very desirable and the internal one is: we just need to do it right. This is an API format that may still be with us in 5-10 years.

It'll still be smaller and more efficient just not as pretty.

> > I do like the idea of transforming the old JSON files into the expected format but at the same time that just seems like moving the compatibility code somewhere else in the codebase.
> 
> If it's in a single location that's easier to delete/remove/deprecate later: that's better than having it inline.
> 
> See the OS layer as an example: it makes it much easier than just having `if OS.mac?` scattered all over the place.

That makes sense.

Reply:
> My bad. I was talking about the idea of having a reduced info cask JSON that could only be used to uninstall casks.

Gotcha. You've convinced me, thanks for taking the time to do so ❤️ 

> The cask JSON doesn't support all of the features found in the cask Ruby file right now so I don't think we'll be able to do that unless we change how the Ruby file works.

What regressions have we caused to users through not writing cask files to the Caskroom? If there are (m)any: have there been (m)any user reports about this (given that most users have been using this for over a year)?

(implication being: few/zero user complaints about this means that this is not actually a meaningful problem for users worth blocking moving everyone to this path)

Reply:
> What regressions have we caused to users through not writing cask files to the Caskroom? If there are (m)any: have there been (m)any user reports about this (given that most users have been using this for over a year)?
>
> (implication being: few/zero user complaints about this means that this is not actually a meaningful problem for users worth blocking moving everyone to this path)

This is related to the casks which are `Cask::Cask#caskfile_only?`. They either have languages or *flight blocks. It hasn't caused any problems because we currently sidestep the problem completely and just download the original cask file (the Ruby one) whenever we install them.

- https://github.com/Homebrew/brew/pull/14820
- https://github.com/Homebrew/brew/pull/14304
- https://github.com/Homebrew/brew/pull/14456

Reply:
Ugh, forgot about that. 

We should probably figure out how to add the necessary data to JSON v3.

I know, scope creep, sorry.